test/extension/handlers/lifecycle/handlers.go

```diff
@@ -0,0 +1,180 @@
+func (h *Handler) DoAfterControlPlaneUpgrade(ctx context.Context, request *runtimehooksv1.AfterControlPlaneUpgradeRequest, response *runtimehooksv1.AfterControlPlaneUpgradeResponse) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("AfterControlPlaneUpgrade is called")
+	cluster := request.Cluster
+	if err := h.recordCallInConfigMap(cluster.Name, cluster.Namespace, runtimehooksv1.AfterControlPlaneUpgrade); err != nil {
+		response.Status = runtimehooksv1.ResponseStatusFailure
+		response.Message = err.Error()
+		return
+	}
+	err := h.readResponseFromConfigMap(cluster.Name, cluster.Namespace, runtimehooksv1.AfterControlPlaneUpgrade, response)
+	if err != nil {
+		response.Status = runtimehooksv1.ResponseStatusFailure
+		response.Message = err.Error()
+		return
+	}
+	log.Info(fmt.Sprintf("AfterControlPlaneUpgrade responding RetryAfterSeconds: %v\n", response.RetryAfterSeconds))
+}
+
+// DoAfterClusterUpgrade implements the AfterClusterUpgrade hook.
+func (h *Handler) DoAfterClusterUpgrade(ctx context.Context, request *runtimehooksv1.AfterClusterUpgradeRequest, response *runtimehooksv1.AfterClusterUpgradeResponse) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("AfterClusterUpgrade is called")
+	cluster := request.Cluster
+	if err := h.recordCallInConfigMap(cluster.Name, cluster.Namespace, runtimehooksv1.AfterClusterUpgrade); err != nil {
+		response.Status = runtimehooksv1.ResponseStatusFailure
+		response.Message = err.Error()
+		return
+	}
+	err := h.readResponseFromConfigMap(cluster.Name, cluster.Namespace, runtimehooksv1.AfterClusterUpgrade, response)
+	if err != nil {
+		response.Status = runtimehooksv1.ResponseStatusFailure
+		response.Message = err.Error()
+		return
+	}
+}
+
+func (h *Handler) readResponseFromConfigMap(name, namespace string, hook runtimecatalog.Hook, response runtimehooksv1.ResponseObject) error {
+	hookName := runtimecatalog.HookName(hook)
+	configMap := &corev1.ConfigMap{}
+	configMapName := name + "-hookresponses"
+	if err := h.Client.Get(context.Background(), client.ObjectKey{Namespace: namespace, Name: configMapName}, configMap); err != nil {
+		return errors.Wrapf(err, "failed to read the ConfigMap %s/%s", namespace, configMapName)
+	}
+	m := map[string]string{}
+	if err := yaml.Unmarshal([]byte(configMap.Data[hookName+"-response"]), m); err != nil {
+		return errors.Wrapf(err, "failed to read %q response information from ConfigMap", hook)
+	}
+	// If the response was a success return early.
+	if status, ok := m["Status"]; ok && status == string(runtimehooksv1.ResponseStatusSuccess) {
+		response.SetStatus(runtimehooksv1.ResponseStatusSuccess)
+		return nil
+	}
+	if _, ok := m["Message"]; ok {
+		response.SetMessage(m["Message"])
+	}
+	if retryResponse, ok := response.(runtimehooksv1.RetryResponseObject); ok {
+		retryAfterSeconds, err := strconv.Atoi(m["RetryAfterSeconds"]) //nolint:gosec
+		if err != nil {
+			return err
+		}
+		retryResponse.SetRetryAfterSeconds(int32(retryAfterSeconds))
+	}
+	return nil
```

Comment:
I think this should not be necessary if you yaml.Unmarshal directly into the response struct. Can you please try?